### PR TITLE
Fix query string handling; clean up tests

### PR DESCRIPTION
**Pull request by timdp**
Improves on #26:

- Appends `URL#search` so we don't lose the query string.
- Extracts the creation of undici options to a testable unit and adds a test for it.
- Checks `Pool#closed` before calling `close`. Without this, I was getting _ClientDestroyedError: The client is destroyed_, so something like a double free. 
- Moves the last assertion of the simple POST test until after the response has been received. By adding some logging, I noticed that the teardown logic was being called before the response had gone through, which seemed shady.
- Replaces `promisifyServerClose` with a simpler `closeServer` helper.

### CODE REVIEW

**Comment by timdp** [comment]
Out of scope: you may want to add `husky` + `lint-staged` to auto-run `standard` on commit.

File: src/fetch.js [hunk 1]
```diff
@@ -3,52 +3,47 @@
...
 
   function fetch (resource, init = {}) {
     const request = new Request(resource, init)
 
-    const origin = request.url.origin
-    let client = clientMap.get(origin)
-    if (client === undefined) {
+    const { origin } = request.url
+    let client
+    if (clientMap.has(origin)) {
```
Comment (Ethan-Arrowood) [inline src/fetch.js:24]: Revert this change, the existing code only has to perform a map lookup once

**Review by Ethan-Arrowood — CHANGES_REQUESTED** [review]
looking good

File: test/fetch.js [hunk 2]
```diff
@@ -41,45 +42,52 @@ tap.test('fetch can handle basic requests', t => {
...
     t.tearDown(async () => {
       await fetch.close()
-      await promisifyServerClose(server)()
+      await closeServer(server)
     })
 
-    server.listen(0, () => {
-      fetch(
-        `http://localhost:${server.address().port}`,
+    server.listen(0, async () => {
```
[the remark this answers is not in the document]
Comment (timdp) [reply test/fetch.js:80]: Certainly true, but it's already implemented like this on main so I thought it wasn't a concern. In other projects, I just promisify `listen`.

File: src/fetch.js [hunk 3]
```diff
@@ -3,52 +3,47 @@
 const Undici = require('undici')
 const Request = require('./request')
 const Response = require('./response')
+const { createUndiciRequestOptions } = require('./utils')
 const { STATUS_CODES } = require('http')
+const { promisify } = require('util')
+
+const requestAsync = promisify(Undici.Pool.prototype.request)
```
[the remark this answers is not in the document]
Comment (timdp) [reply src/fetch.js:10]: Ah, indeed. On main, it gets a callback so I assumed it was CPS only. I'll change it.

File: src/utils.js [hunk 4]
```diff
@@ -69,6 +69,22 @@ function normalizeAndValidateHeaderArguments (name, value) {
   return [normalizeAndValidateHeaderName(name), normalizeAndValidateHeaderValue(name, value)]
 }
 
+function createUndiciRequestOptions (request, signal) {
```
[the remark this answers is not in the document]
Comment (timdp) [reply src/utils.js:72]: I wanted to make it more DRY, but I don't want to bikeshed about destructuring so sure, I'll change it back.

File: src/fetch.js (hunk 1 above)
Comment (timdp) [reply src/fetch.js:24]: I was always told that decent collections perform better with recently looked up values, and a `has` immediately followed by a `get` of the same key is therefore equally fast, while making the code a little more human-readable. But for lack of a benchmark, sure, I'll revert it.

File: test/fetch.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ethan-Arrowood) [reply test/fetch.js:80]: > it's already implemented like this on main so I thought it wasn't a concern

where is this instance? I'd like to clean that up if its in this repo somewhere.

> In other projects, I just promisify listen.

You are probably introducing unknown memory leaks doing this. The `listen` callback method is actually an event listener, and so unless you are using the `on` or `once` methods from `events` promisifying `listen` isn't a good idea.

Honestly, I used to think the same until recently when I started digging into it more over on the main undici library.

File: src/fetch.js (hunk 1 above)
Comment (Ethan-Arrowood) [reply src/fetch.js:24]: Yeah you're probably right, and honestly its going to be very minimal perf difference, but my extra rule of thumb is if its trivially easy to avoid extra method calls then I do so; thus I never even have to concern myself with "decent collections perform better with recently looked up values".

File: test/fetch.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (timdp) [reply test/fetch.js:80]: > where is this instance? I'd like to clean that up if its in this repo somewhere.

https://github.com/Ethan-Arrowood/undici-fetch/blob/ae6d64e830c8d3f6880f5f3423694c4695cb0999/test/fetch.js#L47
https://github.com/Ethan-Arrowood/undici-fetch/blob/ae6d64e830c8d3f6880f5f3423694c4695cb0999/test/fetch.js#L104
https://github.com/Ethan-Arrowood/undici-fetch/blob/ae6d64e830c8d3f6880f5f3423694c4695cb0999/test/fetch.js#L142
https://github.com/Ethan-Arrowood/undici-fetch/blob/ae6d64e830c8d3f6880f5f3423694c4695cb0999/test/fetch.js#L169
https://github.com/Ethan-Arrowood/undici-fetch/blob/ae6d64e830c8d3f6880f5f3423694c4695cb0999/test/fetch.js#L214

> You are probably introducing unknown memory leaks doing this.

When I said "promisify", I meant I wrap the callback in a one-off promise, not that I use the generic `promisify` helper for CPS. That replaces the callback cruft and lets you use `await` without a memory leak.

All five async listeners above can be replaced with such a helper. I'm not super familiar with `tap` but I would expect the tests to look something like:

```js
tap.test('request', t => {
  t.test('handles GET', async t => {
    const serverAdapter = await startServer({
      // routes and such, e.g.,
      '/path': (req, res) => {
        if (req.method === 'GET') {
          res.send('body')
        } else {
          res.status(404)
        }
      }
    })
    t.tearDown(async () => {
      await serverAdapter.close()
    })
    const res = await fetch($server.url + '/path')
    t.strictEqual(await res.text(), 'body')
  })
})
```

but that's _way_ out of scope for this PR. I noticed you've worked on fastify so why not use that in the tests?

File: src/fetch.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (timdp) [reply src/fetch.js]: be6f3fe

File: src/fetch.js [hunk 5]
```diff
@@ -3,52 +3,47 @@
...
-  }
+    const requestOptions = createUndiciRequestOptions(request, init.signal)
 
-  fetch.close = () => {
-    const clientClosePromises = []
-    for (const [, client] of clientMap) {
-      clientClosePromises.push(client.close.bind(client)())
-    }
-    return Promise.all(clientClosePromises)
+    return requestAsync.call(client, requestOptions)
```
[the remark this answers is not in the document]
Comment (timdp) [reply src/fetch.js]: be6f3fe

File: src/utils.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (timdp) [reply src/utils.js:72]: 08a1781

File: src/fetch.js (hunk 1 above)
Comment (timdp) [reply src/fetch.js]: a5e0c62

**Comment by timdp** [comment]
Let me know if you want me to squash. (I'm a fan of squash-merging, personally.)

File: test/fetch.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ethan-Arrowood) [reply test/fetch.js:80]: No point loading all of fastify just for some http client tests. node's core http server should be enough. But yes it seems like theres some improvements to do here. Let's document these changes in a new issue and get this PR wrapped up without too much scope creep. We can follow up with another PR improving the tests later.

**Review by Ethan-Arrowood — CHANGES_REQUESTED** [review]

**Comment by Ethan-Arrowood** [comment]
> Let me know if you want me to squash. (I'm a fan of squash-merging, personally.)

I'll squash on merge

File: src/fetch.js [hunk 6]
```diff
@@ -3,52 +3,42 @@
...
-    for (const [, client] of clientMap) {
-      clientClosePromises.push(client.close.bind(client)())
-    }
-    return Promise.all(clientClosePromises)
+    return client.request(requestOptions)
+      .then(data => new Response(data.body, {
+        status: data.statusCode,
+        statusText: STATUS_CODES[data.statusCode],
+        headers: data.headers
+      }))
```
[the remark this answers is not in the document]
Comment (timdp) [reply src/fetch.js:33]: Why? Rejections propagate so it's redundant.

**Comment by timdp** [comment]
If size is a concern, you could use `micro`. But personally, I think stability and readability are more important for tests. It's just a dev dependency.

File: src/fetch.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ethan-Arrowood) [reply src/fetch.js:33]: I usually do so since its clearer to future collaborators + if there is anything to be cleaned up on error the catch block will be where to do it.

[the remark this answers is not in the document]
Comment (timdp) [reply src/fetch.js:33]: I don't see the point but I don't want to get into a discussion about checked exceptions, so I'll just change it.

However, maybe the broader question is whether we should be forwarding exceptions from undici. They're not compatible with the Fetch spec, as exemplified by #30. So I'll also add a todo.

[the remark this answers is not in the document]
Comment (timdp) [reply src/fetch.js:33]: d1d0b0d

File: test/fetch.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (timdp) [reply test/fetch.js:80]: If size is a concern, you could use `micro`. But personally, I think stability and readability are more important for tests. It's just a dev dependency.

**Review by Ethan-Arrowood — APPROVED** [review]
great work! And yes about the errors. It should be something we fix before v1

**Comment by timdp** [comment]
Cool. Thanks for the reviews!
